corehq/blobs/migrate.py

```diff
@@ -0,0 +1,217 @@
+./manage.py migrate
+
+Note: --file=FILE is optional and can be omitted if you do not want to
+keep a copy of the pre-migrated couch documents.
+
+See also:
+https://github.com/dimagi/commcare-hq/blob/master/corehq/blobs/migrate.py
+"""
+
+BLOB_DB_NOT_CONFIGURED = """
+Cannot get the blob database.
+
+This often means that settings.SHARED_DRIVE_ROOT is not configured.
+It should be set to a real directory. Update localsettings.py and
+retry the migration.
+"""
+
+
+class Migrator(object):
+
+    def __init__(self, slug, doc_types):
+        self.slug = slug
+        self.doc_types = doc_types
+
+    def migrate(self, filename=None):
+        return migrate(self.slug, self.doc_types, filename)
+
+
+MIGRATIONS = {m.slug: m for m in [
+    Migrator("saved_exports", [SavedBasicExport]),
+]}
+
+
+def migrate(slug, doc_types, filename=None):
+    """Migrate attachments from couchdb to blob storage
+
+    The blob storage backend is associated with the model class.
+
+    :param doc_types: List of couch model classes with attachments to be migrated.
+    :param filename: File path for intermediate storage of migration data.
+    """
+    couchdb = doc_types[0].get_db()
+    assert all(t.get_db() is couchdb for t in doc_types[1:]), repr(doc_types)
+    type_map = {cls.__name__: cls for cls in doc_types}
+
+    dirpath = None
+    if filename is None:
+        dirpath = mkdtemp()
+        filename = os.path.join(dirpath, "export.txt")
+
+    print("Loading documents: {}...".format(", ".join(type_map)))
+    total = 0
+    with open(filename, 'w') as f:
+        for doc in get_all_docs_with_doc_types(couchdb, list(type_map)):
+            if doc.get("_attachments"):
+                f.write('{}\n'.format(json.dumps(doc)))
+                total += 1
+
+    with IterDB(couchdb) as iter_db, open(filename, 'r') as f:
+        for n, line in enumerate(f):
+            if n % 100 == 0:
+                print_status(n + 1, total)
+            doc = json.loads(line)
+            obj = type_map[doc["doc_type"]](doc)
+            for name, meta in obj._attachments.iteritems():
+                content = StringIO(meta["data"].decode("base64"))
+                obj.put_attachment(
+                    content, name, content_type=meta["content_type"])
+
+            # delete couch attachments - http://stackoverflow.com/a/2750476
+            obj._attachments.clear()
+
+            iter_db.save(obj.to_json())
+
+    if dirpath is not None:
+        os.remove(filename)
+        os.rmdir(dirpath)
+
+    BlobMigrationState.objects.get_or_create(slug=slug)[0].save()
+    print("Migrated {} documents with attachments. Done.".format(total))
+    return total
+
+
+def print_status(num, total):
+    print("Migrating {} of {} documents with attachments".format(num, total))
+
+
+def assert_migration_complete(slug):
+
+    def forwards(apps, schema_editor):
+        if settings.UNIT_TESTING or getattr(settings, 'IS_TRAVIS', False):
```

Comment:
It's cause I had to run migrate on Travis in order to setup bower testing
which hits a view.  In that case unit_testing is false and so the doc type
migration breaks the migrate command. Maybe a comment would help

On Tuesday, December 1, 2015, Daniel Miller <notifications@github.com>
wrote:

> In corehq/blobs/migrate.py
> <https://github.com/dimagi/commcare-hq/pull/9287#discussion_r46322775>:
>
> > +        os.remove(filename)
> > +        os.rmdir(dirpath)
> > +
> > +    BlobMigrationState.objects.get_or_create(slug=slug)[0].save()
> > +    print("Migrated {} documents with attachments. Done.".format(total))
> > +    return total
> > +
> > +
> > +def print_status(num, total):
> > +    print("Migrating {} of {} documents with attachments".format(num, total))
> > +
> > +
> > +def assert_migration_complete(slug):
> > +
> > +    def forwards(apps, schema_editor):
> > +        if settings.UNIT_TESTING or getattr(settings, 'IS_TRAVIS', False):
>
> Hmm, interesting that he needed that since settings.UNIT_TESTING should
> presumably be set when travis is running tests. I'll remove it here.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dimagi/commcare-hq/pull/9287/files#r46322775>.
>